### PR TITLE
Memory leak in http command.

### PR DESCRIPTION
**Pull request by divolgin**

### CODE REVIEW

**Comment by wcummings** [comment]
Sorry for the delay here, I'll take a look at this tonight

File: c_src/cb.c [hunk 1]
```diff
@@ -651,6 +655,8 @@ ERL_NIF_TERM cb_http(ErlNifEnv* env, handle_t* handle, void* obj)
     ErlNifBinary value_binary;
     enif_alloc_binary(cb.ret.size, &value_binary);
     memcpy(value_binary.data, cb.ret.data, cb.ret.size);
+    free(cb.ret.data);
+    free(cb.ret.key);
```
Comment (wcummings) [inline c_src/cb.c]: Where is `key` being allocated?

Comment (wcummings) [inline c_src/cb.c]: Where is `key` allocated?

**Comment by wcummings** [comment]
The last `free` might be frivolous, LGTM once thats fixed

File: c_src/cb.c (hunk 1 above)
Comment (divolgin) [inline c_src/cb.c]: It appears to be allocated in lcb_make_http_request along with `data`.  In my case, a major leak was caused by the `data` part.  The other four pieces of data didn't contribute that much.

**Comment by divolgin** [comment]
It does leak there too.  Would something like this be acceptable?

```c
// before lcb_make_http_request
cb.ret.key = NULL;
```
...

```c
// at the end
if (cb.ret.key) free(cb.ret.key);
```

File: c_src/cb.c (hunk 1 above)
Comment (wcummings) [inline c_src/cb.c]: If I'm not mistaken, `data` and `key` are allocated by callbacks in callbacks.c for some functions. https://github.com/chitika/cberl/blob/master/c_src/callbacks.c#L111

`data` is allocated but `key` is not, in this case.

**Comment by fauxsoup** [comment]
As long as we initialize both `data` and `key` to NULL, no check should be required before calling free. C specification indicates free should do nothing when passed a NULL pointer, per ISO-IEC 9899.

**Comment by divolgin** [comment]
@fauxsoup fair enough.  I've added explicit initialization and updated the commit.

**Comment by wcummings** [comment]
`key` isn't allocated in the callback, so there's nothing to free. Is there another reason to believe `key` is being leaked here?

File: c_src/cb.c (hunk 1 above)
Comment (divolgin) [inline c_src/cb.c]: It's a pointer.  Any reason not to free it?  The callback code can change.

**Comment by divolgin** [comment]
It's a pointer.  Any reason not to free it?  The callback code can change.

**Comment by wcummings** [comment]
Fair enough. That being the case, we might want to think about abstracting this cleanup logic so that its handled for all functions, not just http. But I can handle that in another PR.

This LGTM :+1:

**Comment by divolgin** [comment]
I agree.  Initially thought about adding another table like [this](https://github.com/chitika/cberl/blob/master/c_src/cberl.h#L19-L20) and calling it `cleanup_calltable`, but it turned out that all other functions already do their own cleanup, this one was the only exception.  If you don't object to that approach, I could spend some time on it at some point.  But I needed this leak fixed quickly for a release.

**Comment by wcummings** [comment]
@divolgin agreed, this can be thought of as a patch fix. I was thinking about something more or less along the lines of `cleanup_calltable`, I'll take a look at it tonight and give it some thought

**Pull request merged**